### ISSUE TITLE
Possible to run tests via only plan instead of plan + apply

### ISSUE THREAD

**Issue by lazarillo** [cloud-custodian/pytest-terraform#38]
`pytest-terraform` launches actual resources via `apply`.  We are planning on testing many resources in many different configurations.  The `plan` portion of the testing takes a few seconds or less, whereas the `apply` can take several minutes each.  Plus, it creates resources that are then destroyed, needlessly creating work.

1. Is there any way in `pytest-terraform` to only run the `plan` and not the `plan`+`apply`?
2. If not, would this be very challenging to implement in the code?
    - (In other words, would it be feasible for me to look through the code and create a pull request capturing this improvement?)
3. Using just `apply`, do you believe that the test will be viable and accurate?  In other words, do you believe that running just `plan` could give accurate results as to whether the deployment will be a success if the plan says that it is a success?

Thank you!  👍

**Comment by kapilt** [comment]
interesting use case, so you instead of terraform resources, you would want the json plan as part of the fixture data structure. i think we could support that use case via some parameterization to the decorator. it also saves the cost/time on teardown.

**Comment by lazarillo** [comment]
Yes, that is what I was hoping.  But I don't think it would be just the JSON plan, would it?  Wouldn't it also need to include a reading of the tfstate file, and possibly writing back to it?

The idea is: `pytest_terraform` would perform all of the proper IaC testing and give all of the valid results, but without having to launch the services.  I understand this requires the assumption that terraform itself can mimic deployment entirely through the call to its plan.  But this is the guarantee that terraform makes, so I figure this should be a safe assumption.

**Comment by kapilt** [comment]
i think you've missed some of the builtin capabilities already extant in pytest-terraform. pytest-terraform will default to recording state files, and can then replay from the disk without doing any api interactions or invocation of terraform, ie record and replay. this uses state file data which is subject to change, versus plan data which can be more incomplete / partial. we have some docs for doing this already in https://cloudcustodian.io/docs/developer/tests.html#writing-tests-for-cloud-controlled-resources we probably need to update that in the pytest-terraform readme.

**Comment by kapilt** [comment]
re last, i just wanted to clarify on the offline capabilities, i do think there's quite a few valid use cases for plan only, which is api interactive, and a more stable api then a state file. in terms of sketching it out how about a parameter to the decorator `@terraform('module_dir', apply=False)` which automatically implies replay=False (although i could see just operating on recorded plans as well). could you clarify if you would want to operate in replay against plans as state?

**Comment by lazarillo** [comment]
@kapilt I have to admit a little ignorance: I am fairly familiar with `pytest`, but not yet with `terraform`.  Moreover, while I have used `pytest` quite a bit, I have never created plugins for it. I looked through the `pytest-terraform` source code and tried to implement in my fork of the repo what I thought would work with my own understanding of how the plugin is working, and I had trouble.

I do not really understand what replay is doing.  I know that Terraform first creates a plan, then applies that plan, and that somewhere in this process the state file is examined to understand what really needs to be altered.  I would say that I am a beginner-to-intermediate terraform *user*, but I have never looked under the hood to see how Terraform works.

As such, I am not really sure how it should be done, or what the best way to handle it is.  Instead, I will just give my use case, from a user perspective:

1. We are establishing IaC on a very large deployment.  Moreover, it is a deployment that is can be automatically rendered.  For an idea of what we are doing, imagine that we are trying to re-create Azure ML workspaces or AWS SageMaker.  It isn't exactly that, but this gives an idea:  we have several interoperating IaC elements that represent ephemeral resources that we want to be able to easily and repeatedly redeploy based upon user needs.
2. We need the robustness that IaC testing will provide (along with unit testing and end-to-end testing of the code deployed on the resources).
3. The team is far more comfortable with Python than with Go, so we would prefer to use `pytest` instead of `terratest` if we can manage it.
4. We need to be able to run many tests in parallel (actually, using async/concurrency concepts rather than parallel processes)
5. We are running so many tests that actually provisioning the resources and waiting for them to respond that they are alive is both very slow, but also it is an unnecessary expense.

I am not sure how to handle all of this yet, and we do not need to solve it all next week or anything, but we are currently in the evaluation phase, and `pytest-terraform` looked promising, but was missing some features we believe will be key.

I am happy to help contribute to code, but I would probably be the only person on my team able to dedicate some time.  As I mentioned, I am not an expert with terraform.

**Comment by kapilt** [comment]
pytest-terraform doesn't need to apply to actually run things you can run in offline mode with recorded outputs (extracted from terraform's state file). so its unclear to me if your trying to provision the resources, just plan for the provisioning of the resources, or just want the rest of the tests using the fixtures to just operate on a recorded run of what had previously been provisioned. the issue with plan typically for consumption by other tests is thats its a delta op set instead of a record of the actual resources, there are valid use cases for both though.

re 4. pytest parallel will do concurrent/parallel processes for executing using pytest-xdist to bring up different fixtures up in parallel, and does maintain that terraform fixtures the notion of per test run execution vs pytest's default per process session scope. by async i assume your referencing an asyncio context which is a programming construct not a functional goal. the underlying cloud apis have fairly low rate limits on writes for control plane infra, so its unclear what your goal is wrt to 4, pytest-terraform will already provision and run tests in parallel.

**Comment by lazarillo** [comment]
Hi @kapilt 

Regarding the first paragraph, all of your proposals sound like great options, but our primary need would be the second, "just plan for the provisioning of the resources".

Regarding the second paragraph, I don't want to get caught up in semantics, but I use "parallel" to mean that multiple tasks can be run at the same time (typically driven by the # of cores, but with `subprocess` it can go beyond that).  Whereas I say "concurrent" to mean that tasks can be "put on hold" while waiting for external resources.  The perfect example for concurrency is web apps:  an endpoint might make a call to a DB that takes 2 seconds to respond.  Using concurrency, the compute resource puts that task "on pause" until it is told of completion, so that the resource can be used for other tasks.  This allows for thousands of tasks to be run concurrently, since the compute requirement is tiny, but the external task is long.

As far as I understand, `pytest-xdist` only performs the parallel paradigm, but not the concurrency paradigm.  Given that any resource provisioning will take a *very* long time, it would be great to use the concurrency paradigm instead.  Of course, many resources depend upon other resources, so there is some inevitable waiting.  But I was hoping to find a package that could run as many tests concurrently as possible, recognizing where dependencies exist.

Thanks!

**Comment by djsd123** [comment]
> re last, i just wanted to clarify on the offline capabilities, i do think there's quite a few valid use cases for plan only, which is api interactive, and a more stable api then a state file. in terms of sketching it out how about a parameter to the decorator `@terraform('module_dir', apply=False)` which automatically implies replay=False (although i could see just operating on recorded plans as well). could you clarify if you would want to operate in replay against plans as state?

@kapilt, Can pytest-terraform be run on the plan json only?  I couldn’t tell from reading the docs.

**Comment by kapilt** [comment]
pytest-terraform can run offline. its not using plan output, its using a saved terraform state file output. i've updated the readme about how to this works, re --replay and record flags. tf_resources.json is automatically written out when running a live test/record, and then --replay can be passed or configured on individual decorators to switch them into running tests with saved output.
